src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_StringIds.cs

```diff
@@ -53,6 +59,14 @@ internal partial class SQLitePersistentStorage
                 return stringId;
             }
 
+            // If we're in a context where our caller doesn't have the write lock, give up now.  They will
+            // call back in with the write lock to allow safe adding of this db ID after this.
+            if (!allowWrite)
+                return null;
+
+            // We're writing.  This better always be under the exclusive scheduler.
+            Contract.ThrowIfFalse(TaskScheduler.Current == _connectionPoolService.Scheduler.ExclusiveScheduler);
```

Comment:
note: techncially allowWrite is redundant.  we could determine this entirely from checking what schduler we're on.  However that made me feel a bit oogy, so i prefer to pass the value in directly.